### PR TITLE
feat: Unique Constraint Violation error translator for different drivers

### PR DESCRIPTION
**Pull request by saeidee**
- [x] Do only one thing
- [x] Non breaking API changes
- [x] Tested

### What did this pull request do?
It will enable developers to have a better error response when there is a unique key constraint violation detected in different SQL engines.

Addresses these issues: https://github.com/go-gorm/gorm/issues/4037, https://github.com/go-gorm/gorm/issues/4135, https://github.com/go-gorm/gorm/issues/5651,  https://github.com/go-gorm/gorm/issues/6002. 

There is no new import or any dependency on any driver, the implementation works by marshaling the returned error from any query and comparing it to SQL engine's default error message by looking after the error code.

```go
err = DB.Create(&City{Name: "Madrid"}).Error
if errors.Is(err, gorm.ErrDuplicatedKey) {
    // some logic here
}
```

### User Case Description

When there is a unique key constraint violation error it will return `gorm.ErrDuplicatedKey`. Basically, it will look if there is a translation available for a specific type of error, if there was one it will return that translation, otherwise, it will return the given untranslated error.

### CODE REVIEW

**Comment by a631807682** [comment]
That's nice, but we'd prefer it to be a [plugin](https://gorm.io/docs/write_plugins.html#Plugin) , or implemented per [drivers](https://github.com/orgs/go-gorm/repositories?q=driver&type=all&language=&sort=) rather than in gorm.

**Comment by jinzhu** [comment]
Maybe we can add `TranslateError` support to dialector.

If the dialector implements `TranslateError` support, will use it to translate the error when `AddError` for db.

**Comment by saeidee** [comment]
> Maybe we can add `TranslateError` support to dialector.
> 
> If the dialector implements `TranslateError` support, will use it to translate the error when `AddError` for db.

This is a good idea, I will update the PR & also create a new PR for one of the drivers.

**Review by a631807682 — APPROVED** [review]

File: gorm.go [hunk 1]
```diff
@@ -347,6 +347,10 @@ func (db *DB) Callback() *callbacks {
 
 // AddError add error to db
 func (db *DB) AddError(err error) error {
+	if errTranslator, ok := db.Dialector.(ErrorTranslator); ok {
```
Comment (a631807682) [inline gorm.go:350]: We don't need to convert when `err == nil`, are you interested in creating an other PR for it?

Comment (saeidee) [reply gorm.go:350]: Yes sure

**Comment by ZhouXing19** [comment]
Hello, will this new `ErrDuplicatedKey` shows up in the next patch (i.e. 1.24.7)?  Thanks!

**Comment by guillermo-st** [comment]
> Hello, will this new `ErrDuplicatedKey` appear in the next patch (i.e. 1.24.7)? Thanks!

I would like to know too! And while I'm at it, how should I go about updating my project dependencies to use the new gorm release that has this error included?

**Comment by saeidee** [comment]
> > Hello, will this new `ErrDuplicatedKey` appear in the next patch (i.e. 1.24.7)? Thanks!
> 
> I would like to know too! And while I'm at it, how should I go about updating my project dependencies to use the new gorm release that has this error included?

Documentation has been added https://github.com/go-gorm/gorm.io/pull/664

**Pull request merged**